File: app/src/main/java/com/jyotirmayg/mytestproject/data/db/AppDatabase.java
package com.jyotirmayg.mytestproject.data.db;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.jyotirmayg.mytestproject.data.db.dao.ItemDao;
import com.jyotirmayg.mytestproject.data.db.entities.Item;

/**
 * @author jyoti
 * @created on 14-05-2022
 * This class is responsible for creating room database.
 */
@Database(entities = Item.class, exportSchema = false, version = 1)
public abstract class AppDatabase extends RoomDatabase {

    private static final String DB_NAME = "item_db";
    private static AppDatabase instance;

    public static synchronized AppDatabase getInstance(Context context) {
        if (instance == null) {
            instance = Room.databaseBuilder(context.getApplicationContext(), AppDatabase.class, DB_NAME)
                    .fallbackToDestructiveMigration().build();
        }
        return instance;
    }

    public abstract ItemDao itemDao();
}

File: app/src/main/java/com/jyotirmayg/mytestproject/data/db/AppDatabase.kt
package com.jyotirmayg.mytestproject.data.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.jyotirmayg.mytestproject.data.db.dao.ItemDao
import com.jyotirmayg.mytestproject.data.db.entities.Item

/**
 * @author jyoti
 * @created on 14-05-2022
 * This class is responsible for creating room database.
 */
@Database(entities = [Item::class], version = 1)
abstract class AppDatabase : RoomDatabase() {

    abstract fun itemDao(): ItemDao

    /**
     * This created a singleton object of the AppDatabase class.
     * And creates the database.
     * */
    companion object {
        @Volatile
        private var instance: AppDatabase? = null
        private val LOCK = Any()

        operator fun invoke(context: Context) = instance ?: synchronized(LOCK) {
            instance ?: buildDatabase(context).also {
                instance = it
            }
        }

        private fun buildDatabase(context: Context) =
            Room.databaseBuilder(
                context.applicationContext,
                AppDatabase::class.java,
                "App.db"
            ).build()
    }
}